File: cracking_the_coding_interview_6th/chapter07/cpp/07_01_v01.cpp
// this version needs dynamic_cast in functions to deal with BJCard.
// I want to refactor that by template, such as Deck<BJCard>.
#include <algorithm>
#include <chrono>
#include <functional>
#include <iostream>
#include <list>
#include <memory>
#include <stdexcept>
#include <string>
#include <type_traits>
#include <vector>
using namespace std;
using ll = long long;

namespace trump {

enum class Suits;
class _Card;

using Card = shared_ptr<_Card>;

using Cards = list<Card>;

using DeckInitializer = function<Cards()>;

enum class Suits { diamond, club, spade, heart };

string suit_str(const Suits &suit) {
  switch (suit) {
  case Suits::diamond:
    return "diamond";
  case Suits::club:
    return "club";
  case Suits::spade:
    return "spade";
  case Suits::heart:
    return "heart";
  }
}

ostream &operator<<(ostream &os, const Suits &suit) {
  os << suit_str(suit);
  return os;
}

const vector<Suits> suits_set{
    Suits::diamond, Suits::club, Suits::spade, Suits::heart};

class _Card {
protected:
  string repr(const vector<string> &args) {
    string res{"Card("};
    for (const auto &s : args) {
      res += s;
      res += ", ";
    }
    res.pop_back();
    res.pop_back();
    return res += ")";
  };

public:
  virtual string to_str() = 0;
};

class OrdinalCard : public _Card {
  Suits _suit;
  ll _rank;
  ll _min_rank = 1;
  ll _max_rank = 13;

public:
  OrdinalCard(Suits suit, ll rank) : _suit{suit} {
    if (rank < _min_rank || _max_rank < rank)
      throw invalid_argument{"_rank must be in [1, 13]"};
    _rank = rank;
  }

  Suits suit() {
    return _suit;
  }

  ll rank() {
    return _rank;
  }

  string to_str() override {
    return repr({suit_str(_suit), to_string(_rank)});
  }
};

class BJCard : public OrdinalCard {
  ll _value;

public:
  BJCard(Suits suit, ll rank) : OrdinalCard{suit, rank} {
    if (rank >= 11)
      _value = 10;
    else
      _value = rank;
  }

  ll value() {
    return _value;
  }

  string to_str() override {
    return repr({suit_str(suit()), to_string(rank()), to_string(_value)});
  }
};

Cards BJ_deck_initializer() {
  Cards cards;
  for (auto s : suits_set) {
    for (ll i = 1; i <= 13; ++i) {
      cards.push_back(make_shared<BJCard>(s, i));
    }
  }
  return cards;
}

class Deck {
  Cards _cards;

public:
  Deck(Cards cards) {}
  Deck(DeckInitializer f) {
    _cards = f();
  }
  Card draw() {
    if (_cards.size() == 0) throw out_of_range{"deck is empty"};
    Card x = _cards.front();
    _cards.pop_front();
    return x;
  }
  void put(Card card) {
    _cards.push_front(card);
  }
  void shuffle() {
    vector<Card> v{_cards.begin(), _cards.end()};
    random_shuffle(v.begin(), v.end());
    _cards = list<Card>(v.begin(), v.end());
  }
  ll size() {
    return _cards.size();
  }
};
} // namespace trump

int main() {
  using namespace trump;
  Deck BJ_deck{BJ_deck_initializer};
  BJ_deck.shuffle();
  Card c1 = BJ_deck.draw();
  cout << c1->to_str() << endl;
  Card c2 = BJ_deck.draw();
  cout << c2->to_str() << endl;
  Card c3 = BJ_deck.draw();
  cout << c3->to_str() << endl;
}

// Card(diamond, 5, 5)
// Card(heart, 13, 10)
// Card(diamond, 12, 10)

File: cracking_the_coding_interview_6th/chapter07/python/07_01_v01.py
from typing import Callable, List, Iterable, Union
from abc import ABC
from enum import Enum, auto
from random import shuffle


class Suits(Enum):
    diamond = auto()
    club = auto()
    spade = auto()
    heart = auto()

    def __str__(self) -> str:
        return self.name


class Card(ABC):
    def repr(self, *args: Iterable[str]) -> str:
        return f"Card({', '.join(args)})"


class OrdinalCard(Card):
    def __init__(self, suit: Suits, rank: int):
        if rank < 0 or 13 < rank:
            raise ValueError("rank must be in [1, 13]")
        self.__suit = suit
        self.__rank = rank

    def __repr__(self) -> str:
        return self.repr(str(self.__suit), str(self.__rank))

    def suit(self) -> Suits:
        return self.__suit

    def rank(self) -> int:
        return self.__rank


class Joker(Card):
    def __repr__(self) -> str:
        return self.repr("Joker")


Cards = List[Card]
DeckInitializer = Callable[[], Cards]


def default_deck_initializer_without_jokers() -> Cards:
    return [OrdinalCard(s, i) for (s, _) in Suits.__members__.items() for i in range(1, 14)]


def default_deck_initializer() -> Cards:
    return default_deck_initializer_without_jokers + [Joker(), Joker()]


class Deck:
    def __init__(self, initializer: Union[Cards, DeckInitializer]):
        if isinstance(initializer, list):
            self.__cards = initializer
        elif callable(initializer):
            self.__cards = initializer()
        else:
            raise ValueError(
                "Invalid argument; initializer must be a Cards or DeckInitializer object")

    def __len__(self) -> int:
        return len(self.__cards)

    def shuffle(self) -> None:
        shuffle(self.__cards)

    def put(self, card: Card) -> None:
        self.__cards.append(card)

    def draw(self) -> Card:
        if len(self) == 0:
            raise ValueError("deck is empty")
        return self.__cards.pop()


def main():
    BJ_deck = Deck(default_deck_initializer_without_jokers)
    BJ_deck.shuffle()
    c1 = BJ_deck.draw()
    print(c1)
    c2 = BJ_deck.draw()
    print(c2)
    c3 = BJ_deck.draw()
    print(c3)

# Card(spade, 7)
# Card(diamond, 10)
# Card(spade, 2)


if __name__ == "__main__":
    main()
